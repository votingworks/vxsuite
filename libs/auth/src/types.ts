/**
 * Valid values for cert card type field
 */
export type CardType =
  | 'system-administrator'
  | 'election-manager'
  | 'poll-worker'
  | 'poll-worker-with-pin';

/**
 * Parsed custom cert fields
 */
export interface CustomCertFields {
  component: 'card' | 'admin' | 'central-scan' | 'mark' | 'scan';
  jurisdiction: string;
  cardType?: CardType;
  electionHash?: string;
}
