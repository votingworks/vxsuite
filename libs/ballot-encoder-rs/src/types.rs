use std::num::NonZeroU8;

use static_assertions::const_assert;

use crate::codable;

codable!(PrecinctIndex, u32, 0..=4096);
codable!(BallotStyleIndex, u32, 0..=4096);
codable!(PageNumber, u8, 1..=30);

// Statically validate our maximum values fit within the types we're using.
// TODO: move this into `codable!`
const_assert!(PrecinctIndex::BITS <= usize::BITS);
const_assert!(BallotStyleIndex::BITS <= usize::BITS);
const_assert!(PageNumber::BITS <= u8::BITS);

impl PageNumber {
    /// Whether this is the first page of its sheet, i.e. the front.
    ///
    /// ```
    /// # use ballot_encoder_rs::types::PageNumber;
    /// let page_one = PageNumber::new_unchecked(1);
    /// let page_two = PageNumber::new_unchecked(2);
    ///
    /// assert!(page_one.is_recto());
    /// assert!(!page_two.is_recto());
    /// ```
    #[must_use]
    pub const fn is_recto(self) -> bool {
        self.0 % 2 == 1
    }

    /// Whether this is the second page of its sheet, i.e. the back.
    ///
    /// ```
    /// # use ballot_encoder_rs::types::PageNumber;
    /// let page_one = PageNumber::new_unchecked(1);
    /// let page_two = PageNumber::new_unchecked(2);
    ///
    /// assert!(!page_one.is_verso());
    /// assert!(page_two.is_verso());
    /// ```
    #[must_use]
    pub const fn is_verso(self) -> bool {
        self.0 % 2 == 0
    }

    /// Determines the sheet number.
    ///
    /// ```
    /// # use ballot_encoder_rs::types::PageNumber;
    /// let page_one = PageNumber::new_unchecked(1);
    /// let page_two = PageNumber::new_unchecked(2);
    /// let page_three = PageNumber::new_unchecked(3);
    /// let page_four = PageNumber::new_unchecked(4);
    ///
    /// assert_eq!(page_one.sheet_number().get(), 1);
    /// assert_eq!(page_two.sheet_number().get(), 1);
    /// assert_eq!(page_three.sheet_number().get(), 2);
    /// assert_eq!(page_four.sheet_number().get(), 2);
    /// ```
    ///
    /// # Panics
    ///
    /// If `Self` is an invalid `PageNumber` that would produce a sheet number
    /// of zero, this method will panic.
    #[must_use]
    pub const fn sheet_number(self) -> NonZeroU8 {
        let verso_page = if self.is_verso() {
            self
        } else {
            self.opposite()
        };
        NonZeroU8::new(verso_page.get() / 2).expect("sheet number >= 1")
    }

    /// Gets the `PageNumber` opposite this one on the same sheet.
    ///
    /// ```
    /// # use ballot_encoder_rs::types::PageNumber;
    /// let page_one = PageNumber::new_unchecked(1);
    /// let page_two = page_one.opposite();
    /// assert_eq!(page_two.get(), 2);
    /// ```
    pub const fn opposite(self) -> Self {
        match self.0 % 2 {
            0 => Self(self.0 - 1),
            1 => Self(self.0 + 1),
            _ => unreachable!(),
        }
    }
}

#[cfg(test)]
pub mod tests {
    use std::num::NonZeroU8;

    use proptest::{prelude::Strategy, proptest};

    use super::PageNumber;

    #[test]
    fn test_page_number() {
        let page_one = PageNumber::new_unchecked(1);
        let page_two = PageNumber::new_unchecked(2);

        assert_eq!(page_one.get(), 1);
        assert_eq!(page_two.get(), 2);

        assert_eq!(page_one, page_one);
        assert_ne!(page_one, page_two);

        assert!(page_one.is_recto());
        assert!(!page_two.is_recto());
        assert!(!page_one.is_verso());
        assert!(page_two.is_verso());

        assert_eq!(page_one.sheet_number(), NonZeroU8::new(1).unwrap());
        assert_eq!(page_two.sheet_number(), NonZeroU8::new(1).unwrap());
    }

    pub fn arbitrary_page_number() -> impl Strategy<Value = PageNumber> {
        (PageNumber::MIN_VALUE..=PageNumber::MAX_VALUE).prop_map(PageNumber::new_unchecked)
    }

    proptest! {
        #[test]
        fn test_valid_page_number(page_number in arbitrary_page_number()) {
            assert_eq!(page_number, page_number.opposite().opposite());
            assert_eq!(page_number.is_recto(), page_number.opposite().is_verso());
            assert_eq!(page_number.is_verso(), page_number.opposite().is_recto());
            assert_eq!(page_number.sheet_number(), page_number.opposite().sheet_number());
        }

        #[test]
        fn test_invalid_page_number(value in PageNumber::MAX_VALUE + 1..=u8::MAX) {
            assert_eq!(PageNumber::new(value), None);
        }
    }
}
