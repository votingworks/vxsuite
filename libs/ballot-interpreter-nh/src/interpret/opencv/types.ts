import * as cv from '@u4/opencv4nodejs';

import {
  GridLayout,
  GridPosition,
  HmpbBallotPageMetadata,
  PageInterpretation,
  Result,
} from '@votingworks/types';

/** TODO */
export interface PageImageData {
  binarizedImage: cv.Mat;
  optimizedImage: cv.Mat;
  originalImage: cv.Mat;
}

/** Single page interpretation result. */
export interface OpenCvInterpretResult {
  interpretation: PageInterpretation;
  normalizedImageFilePath: string;
  originalImageFilePath: string;
}

/** TODO */
export interface PageRegion {
  image: cv.Mat;
  offset: cv.Point2;
}

/** TODO */
export interface PageMargins<T> {
  bottom: T;
  left: T;
  right: T;
  top: T;
}

/** TODO */
export type MarginType = keyof PageMargins<unknown>;

/** TODO */
export interface PotentialTimingMarks {
  allContours: cv.Contour[];
  potentialMarks: cv.Contour[];
}

/** TODO */
export interface DetectedTimingMarks {
  marginImage: PageRegion;
  marginType: MarginType;
  marks: cv.Contour[];
  medianHeight: number;
  medianSpacing: number;
  medianWidth: number;
  // pageRegion: PageRegion;
}

/** TODO */
export interface TimingMarks {
  detected: cv.Contour[];
  first: cv.Rect;
  interpolated: cv.Rect[];
  last: cv.Rect;
  medianSpacing: number;
  marginType: MarginType;
}

/** TODO */
export type BallotMetadata = Pick<
  HmpbBallotPageMetadata,
  'ballotStyleId' | 'ballotType' | 'electionHash' | 'isTestMode' | 'precinctId'
> & {
  gridLayout: GridLayout;
};

/** TODO */
export interface OvalTemplate {
  image: cv.Mat;
  shadedPixelRatio: number;
}

/** TODO */
export interface OvalMark {
  readonly gridPosition: GridPosition;
  readonly expectedBounds: Result<cv.Rect, string>;
  readonly matchBounds: Result<cv.Rect, string>;
  readonly score: number;
  readonly searchBounds: Result<cv.Rect, string>;
}
