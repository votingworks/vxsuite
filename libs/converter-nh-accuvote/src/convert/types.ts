import { Result } from '@votingworks/basics';
import {
  BallotPaperSize,
  Election,
  GridPosition,
  Size,
} from '@votingworks/types';
import { ZodError } from 'zod';
import { PartialTimingMarks } from '@votingworks/ballot-interpreter';
import { ParseConstitutionalQuestionError } from './parse_constitutional_questions';

/**
 * The kinds of errors that can occur during `pairColumnEntries`.
 */
export enum PairColumnEntriesIssueKind {
  ColumnCountMismatch = 'ColumnCountMismatch',
  ColumnEntryCountMismatch = 'ColumnEntryCountMismatch',
}

/**
 * Errors that can occur during `pairColumnEntries`.
 */
export type PairColumnEntriesIssue<T extends GridEntry, U extends GridEntry> =
  | {
      kind: PairColumnEntriesIssueKind.ColumnCountMismatch;
      message: string;
      columnCounts: [number, number];
    }
  | {
      kind: PairColumnEntriesIssueKind.ColumnEntryCountMismatch;
      message: string;
      columnIndex: number;
      columnEntryCounts: [number, number];
      extraLeftEntries: T[];
      extraRightEntries: U[];
    };

/**
 * Result of {@link pairColumnEntries}. The `issues` property is an array of
 * issues that occurred during the pairing process. The `Err` variant still has
 * `pairs`, but they will only be partially populated.
 */
export type PairColumnEntriesResult<
  T extends GridEntry,
  U extends GridEntry,
> = Result<
  {
    readonly pairs: ReadonlyArray<[T, U]>;
  },
  {
    readonly pairs: ReadonlyArray<[T, U]>;
    readonly issues: ReadonlyArray<PairColumnEntriesIssue<T, U>>;
  }
>;

/**
 * Contains the metadata and ballot images for a ballot card.
 */
export interface NewHampshireBallotCardDefinition {
  /**
   * XML element containing the ballot card definition, including election info
   * and contests with candidates.
   */
  readonly definition: Element;

  /**
   * An image of the ballot card's front as rendered from a PDF.
   */
  readonly front: ImageData;

  /**
   * An image of the ballot card's back as rendered from a PDF.
   */
  readonly back: ImageData;
}

/**
 * Contains candidate elements and their LCM column/row coordinates.
 */
export interface CandidateGridElement {
  readonly element: Element;
  readonly column: number;
  readonly row: number;
}

/**
 * Basic properties for an object located on a grid.
 */
export interface GridEntry {
  readonly side: 'front' | 'back';
  readonly column: number;
  readonly row: number;
}

/**
 * Kinds of errors that can occur when converting a ballot card definition.
 */
export enum ConvertIssueKind {
  ElectionValidationFailed = 'ElectionValidationFailed',
  InvalidBallotSize = 'InvalidBallotSize',
  InvalidDistrictId = 'InvalidDistrictId',
  InvalidElectionDate = 'InvalidElectionDate',
  InvalidTemplateSize = 'InvalidTemplateSize',
  InvalidTimingMarkMetadata = 'InvalidTimingMarkMetadata',
  MismatchedBallotImageSize = 'MismatchedBallotImageSize',
  MismatchedOvalGrids = 'MismatchedOvalGrids',
  MissingDefinitionProperty = 'MissingDefinitionProperty',
  MissingTimingMarkMetadata = 'MissingTimingMarkMetadata',
  TimingMarkDetectionFailed = 'TimingMarkDetectionFailed',
  ConstitutionalQuestionError = 'ConstitutionalQuestionError',
}

/**
 * Represents a bubble found in a ballot template.
 */
export interface TemplateBubble {
  readonly row: number;
  readonly column: number;
}

/**
 * A grid entry for a specific template oval.
 */
export type TemplateBubbleGridEntry = TemplateBubble & {
  side: 'front' | 'back';
};

/**
 * Issues that can occur when converting a ballot card definition.
 */
export type ConvertIssue =
  | {
      kind: ConvertIssueKind.ElectionValidationFailed;
      message: string;
      validationError: Error;
    }
  | {
      kind: ConvertIssueKind.InvalidBallotSize;
      message: string;
      invalidBallotSize: string;
    }
  | {
      kind: ConvertIssueKind.InvalidTemplateSize;
      message: string;
      paperSize?: BallotPaperSize;
      frontTemplateSize: Size;
      backTemplateSize: Size;
    }
  | {
      kind: ConvertIssueKind.InvalidDistrictId;
      message: string;
      invalidDistrictId: string;
    }
  | {
      kind: ConvertIssueKind.InvalidElectionDate;
      message: string;
      invalidDate: string;
      invalidReason: string;
    }
  | {
      kind: ConvertIssueKind.InvalidTimingMarkMetadata;
      message: string;
      side: 'front' | 'back';
      timingMarks: PartialTimingMarks;
      timingMarkBits: readonly boolean[];
      validationError?: ZodError;
    }
  | {
      kind: ConvertIssueKind.MismatchedOvalGrids;
      message: string;
      pairColumnEntriesIssue: PairColumnEntriesIssue<
        GridPosition,
        TemplateBubbleGridEntry
      >;
    }
  | {
      kind: ConvertIssueKind.MissingDefinitionProperty;
      message: string;
      property: string;
    }
  | {
      kind: ConvertIssueKind.MissingTimingMarkMetadata;
      message: string;
      side: 'front' | 'back';
      timingMarks: PartialTimingMarks;
    }
  | {
      kind: ConvertIssueKind.TimingMarkDetectionFailed;
      message: string;
      side: 'front' | 'back';
    }
  | {
      kind: ConvertIssueKind.ConstitutionalQuestionError;
      message: string;
      error: ParseConstitutionalQuestionError;
    };

/**
 * Contains the result of converting a ballot card definition.
 */
export type ConvertResult = Result<
  {
    readonly election: Election;
    readonly issues: readonly ConvertIssue[];
  },
  {
    readonly election?: Election;
    readonly issues: readonly ConvertIssue[];
  }
>;
