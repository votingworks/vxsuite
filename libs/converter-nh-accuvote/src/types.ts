import { BallotPaperSize, GridPosition, NewType } from '@votingworks/types';
import { z } from 'zod';

/**
 * A point in 2D space.
 */
export type Point = NewType<
  {
    readonly x: number;
    readonly y: number;
  },
  'Point'
>;

/**
 * Schema for {@link Point}.
 */
export const PointSchema = z.object({
  x: z.number(),
  y: z.number(),
}) as unknown as z.ZodSchema<Point>;

/**
 * Insets from a rectangle 2D space.
 */
export interface Inset {
  readonly left: number;
  readonly top: number;
  readonly right: number;
  readonly bottom: number;
}

/**
 * A vector in 2D space.
 */
export type Vector = NewType<
  {
    readonly x: number;
    readonly y: number;
  },
  'Vector'
>;

/**
 * Schema for {@link Vector}.
 */
export const VectorSchema = z.object({
  x: z.number(),
  y: z.number(),
}) as unknown as z.ZodSchema<Vector>;

/**
 * A size in 2D space.
 */
export interface Size {
  readonly width: number;
  readonly height: number;
}

/**
 * A rectangle in 2D space.
 */
export interface Rect {
  readonly x: number;
  readonly y: number;
  readonly width: number;
  readonly height: number;
  readonly minX: number;
  readonly minY: number;
  readonly maxX: number;
  readonly maxY: number;
}

/**
 * A line segment in 2D space.
 */
export interface Segment {
  readonly from: Point;
  readonly to: Point;
}

/**
 * Potentially-incomplete timing marks for a ballot card.
 */
export interface PartialTimingMarks {
  readonly bottom: readonly Rect[];
  readonly left: readonly Rect[];
  readonly right: readonly Rect[];
  readonly top: readonly Rect[];
  readonly topLeft?: Rect;
  readonly topRight?: Rect;
  readonly bottomLeft?: Rect;
  readonly bottomRight?: Rect;
}

/**
 * Complete timing marks for a ballot card. Includes all timing marks,
 * meaning that the timing marks are evenly spaced and left and right
 * have the same number of timing marks. Also includes the corners. All top
 * timing marks are included, though bottom marks may be missing because the
 * bottom row is used for encoding metadata.
 */
export interface CompleteTimingMarks {
  readonly bottom: readonly Rect[];
  readonly left: readonly Rect[];
  readonly right: readonly Rect[];
  readonly top: readonly Rect[];
  readonly topLeft: Rect;
  readonly topRight: Rect;
  readonly bottomLeft: Rect;
  readonly bottomRight: Rect;
}

/**
 * Contains a grid of all possible locations for option bubbles on the ballot
 * card. If a ballot card has N columns and M rows, there are (N - 1) * (M - 1)
 * potential option bubble locations.
 */
export interface PossibleOptionBubblesGrid {
  readonly rows: ReadonlyArray<readonly Point[]>;
}

/**
 * Represents a single bit, yes = 1, no = 0.
 */
export type Bit = 0 | 1;

/**
 * Represents two bits.
 */
export type TwoBits = readonly [Bit, Bit];

/**
 * Represents four bits.
 */
export type FourBits = readonly [...TwoBits, ...TwoBits];

/**
 * Represents eight bits, or a single byte.
 */
export type EightBits = readonly [...FourBits, ...FourBits];

/**
 * Represents sixteen bits, or two bytes.
 */
export type SixteenBits = readonly [...EightBits, ...EightBits];

/**
 * Represents thirty-two bits, or four bytes.
 */
export type ThirtyTwoBits = readonly [...SixteenBits, ...SixteenBits];

/**
 * Schema for {@link Bit}.
 */
export const BitSchema: z.ZodSchema<Bit> = z.union([
  z.literal(0),
  z.literal(1),
]);

/**
 * Schema for {@link ThirtyTwoBits}.
 */
export const ThirtyTwoBitsSchema: z.ZodSchema<ThirtyTwoBits> = z.tuple([
  BitSchema,
  BitSchema,
  BitSchema,
  BitSchema,
  BitSchema,
  BitSchema,
  BitSchema,
  BitSchema,
  BitSchema,
  BitSchema,
  BitSchema,
  BitSchema,
  BitSchema,
  BitSchema,
  BitSchema,
  BitSchema,
  BitSchema,
  BitSchema,
  BitSchema,
  BitSchema,
  BitSchema,
  BitSchema,
  BitSchema,
  BitSchema,
  BitSchema,
  BitSchema,
  BitSchema,
  BitSchema,
  BitSchema,
  BitSchema,
  BitSchema,
  BitSchema,
]);

/**
 * Metadata encoded by the bottom row of the front of a ballot card.
 */
export interface FrontMarksMetadata {
  /**
   * Discriminator for the type of metadata.
   */
  readonly side: 'front';

  /**
   * Raw bits 0-31 in LSB-MSB order (right-to-left).
   */
  readonly bits: ThirtyTwoBits;

  /**
   * Mod 4 check sum from bits 0-1 (2 bits).
   *
   * The mod 4 check sum bits are obtained by adding the number of 1’s in bits 2
   * through 31, then encoding the results of a mod 4 operation in bits 0 and 1.
   * For example, if bits 2 through 31 have 18 1’s, bits 0 and 1 will hold the
   * value 2 (18 mod 4 = 2).
   */
  readonly mod4CheckSum: number;

  /**
   * The mod 4 check sum computed from bits 2-31.
   */
  readonly computedMod4CheckSum: number;

  /**
   * Batch or precinct number from bits 2-14 (13 bits).
   */
  readonly batchOrPrecinctNumber: number;

  /**
   * Card number (CardRotID) from bits 15-27 (13 bits).
   */
  readonly cardNumber: number;

  /**
   * Sequence number (always 0) from bits 28-30 (3 bits).
   */
  readonly sequenceNumber: number;

  /**
   * Start bit (always 1) from bit 31-31 (1 bit).
   */
  readonly startBit: Bit;
}

/**
 * Schema for {@link FrontMarksMetadata}.
 */
export const FrontMarksMetadataSchema: z.ZodSchema<FrontMarksMetadata> = z
  .object({
    side: z.literal('front'),
    bits: ThirtyTwoBitsSchema,
    mod4CheckSum: z.number(),
    computedMod4CheckSum: z.number(),
    batchOrPrecinctNumber: z.number().int().nonnegative(),
    cardNumber: z.number().int().nonnegative(),
    sequenceNumber: z
      .number()
      .int()
      .nonnegative()
      .refine((n) => n === 0, 'sequenceNumber must be 0'),
    startBit: BitSchema.refine(
      (bit) => bit === 1,
      'startBit must be 1'
    ) as typeof BitSchema,
  })
  .refine(
    ({ mod4CheckSum, computedMod4CheckSum }) =>
      mod4CheckSum === computedMod4CheckSum,
    'mod4CheckSum must equal computedMod4CheckSum'
  );

/**
 * Metadata encoded by the bottom row of the back of a ballot card.
 */
export interface BackMarksMetadata {
  /**
   * Discriminator for the type of metadata.
   */
  readonly side: 'back';

  /**
   * Raw bits 0-31 in LSB-MSB order (right-to-left).
   */
  readonly bits: ThirtyTwoBits;

  /**
   * Election day of month (1..31) from bits 0-4 (5 bits).
   */
  readonly electionDay: number;

  /**
   * Election month (1..12) from bits 5-8 (4 bits).
   */
  readonly electionMonth: number;

  /**
   * Election year (2 digits) from bits 9-15 (7 bits).
   */
  readonly electionYear: number;

  /**
   * Election type from bits 16-20 (5 bits).
   *
   * @example "G" for general election
   */
  readonly electionType:
    | 'A'
    | 'B'
    | 'C'
    | 'D'
    | 'E'
    | 'F'
    | 'G'
    | 'H'
    | 'I'
    | 'J'
    | 'K'
    | 'L'
    | 'M'
    | 'N'
    | 'O'
    | 'P'
    | 'Q'
    | 'R'
    | 'S'
    | 'T'
    | 'U'
    | 'V'
    | 'W'
    | 'X'
    | 'Y'
    | 'Z';

  /**
   * Ender code (binary 01111011110) from bits 21-31 (11 bits).
   */
  readonly enderCode: [Bit, ...TwoBits, ...EightBits];

  /**
   * Ender code (binary 01111011110) hardcoded to the expected value.
   */
  readonly expectedEnderCode: [Bit, ...TwoBits, ...EightBits];
}

/**
 * Schema for {@link BackMarksMetadata}.
 */
export const BackMarksMetadataSchema: z.ZodSchema<BackMarksMetadata> = z
  .object({
    side: z.literal('back'),
    bits: ThirtyTwoBitsSchema,
    electionDay: z
      .number()
      .int()
      .positive()
      .refine((n) => n <= 31, 'electionDay must be between 1 and 31'),
    electionMonth: z
      .number()
      .int()
      .positive()
      .refine((n) => n <= 12, 'electionMonth must be between 1 and 12'),
    electionYear: z
      .number()
      .int()
      .positive()
      .refine((n) => n <= 99, 'electionYear must be between 0 and 99'),
    electionType: z.union([
      z.literal('A'),
      z.literal('B'),
      z.literal('C'),
      z.literal('D'),
      z.literal('E'),
      z.literal('F'),
      z.literal('G'),
      z.literal('H'),
      z.literal('I'),
      z.literal('J'),
      z.literal('K'),
      z.literal('L'),
      z.literal('M'),
      z.literal('N'),
      z.literal('O'),
      z.literal('P'),
      z.literal('Q'),
      z.literal('R'),
      z.literal('S'),
      z.literal('T'),
      z.literal('U'),
      z.literal('V'),
      z.literal('W'),
      z.literal('X'),
      z.literal('Y'),
      z.literal('Z'),
    ]),
    enderCode: z.tuple([
      BitSchema,
      BitSchema,
      BitSchema,
      BitSchema,
      BitSchema,
      BitSchema,
      BitSchema,
      BitSchema,
      BitSchema,
      BitSchema,
      BitSchema,
    ]),
    expectedEnderCode: z.tuple([
      BitSchema,
      BitSchema,
      BitSchema,
      BitSchema,
      BitSchema,
      BitSchema,
      BitSchema,
      BitSchema,
      BitSchema,
      BitSchema,
      BitSchema,
    ]),
  })
  .refine(
    ({ enderCode, expectedEnderCode }) =>
      enderCode.length === expectedEnderCode.length &&
      enderCode.every((bit, i) => bit === expectedEnderCode[i]),
    'enderCode must be the expected value'
  );

/**
 * Metadata encoded by the front or back of a ballot card.
 */
export type MarksMetadata = FrontMarksMetadata | BackMarksMetadata;

/**
 * Schema for {@link MarksMetadata}.
 */
export const MarksMetadataSchema: z.ZodSchema<MarksMetadata> = z.union([
  FrontMarksMetadataSchema,
  BackMarksMetadataSchema,
]);

/**
 * An interpretation of an oval mark on a ballot.
 */
export interface InterpretedOvalMark {
  readonly gridPosition: GridPosition;
  readonly score: number;
  readonly scoredOffset: Vector;
  readonly bounds: Rect;
}

/**
 * Geometry information about an ballot card with timing marks.
 */
export interface BallotCardGeometry {
  /**
   * The size of the ballot card as a standard paper size.
   */
  readonly ballotPaperSize: BallotPaperSize;

  /**
   * Pixels per inch of the ballot card.
   */
  readonly pixelsPerInch: number;

  /**
   * The size of the ballot card in pixels.
   */
  readonly canvasSize: Size;

  /**
   * The content area of the ballot card in pixels.
   */
  readonly contentArea: Rect;

  /**
   * The size of a timing mark in pixels.
   */
  readonly timingMarkSize: Size;

  /**
   * The size of an oval in pixels.
   */
  readonly ovalSize: Size;

  /**
   * The size of the grid of timing marks, in units of timing marks. For
   * example, a grid of size `{ width: 20, height: 30 }` means there are 20
   * columns of timing marks and 30 rows of timing marks.
   */
  readonly gridSize: Size;

  /**
   * The area within the timing mark grid on the front of the ballot card that
   * may be used for ovals. In practice, there will be at least one edge in each
   * direction that is not usable for ovals.
   */
  readonly frontUsableArea: Rect;

  /**
   * The area within the timing mark grid on the back of the ballot card that
   * may be used for ovals. In practice, there will be at least one edge in each
   * direction that is not usable for ovals.
   */
  readonly backUsableArea: Rect;
}

/**
 * Ballot card orientation.
 */
export enum BallotCardOrientation {
  /**
   * The ballot card is portrait and right-side up.
   */
  Portrait = 'portrait',

  /**
   * The ballot card is portrait and upside down.
   */
  PortraitReversed = 'portrait-reversed',

  /**
   * The ballot card is landscape and right-side up.
   */
  Landscape = 'landscape',

  /**
   * The ballot card is landscape and upside down.
   */
  LandscapeReversed = 'landscape-reversed',
}
