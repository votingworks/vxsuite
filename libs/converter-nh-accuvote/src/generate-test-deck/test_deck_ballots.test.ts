import { electionGridLayoutNewHampshireHudsonFixtures } from '@votingworks/fixtures';
import { CandidateId, ContestId, GridPosition } from '@votingworks/types';
import { generateHandMarkedTestDeckBallots } from './test_deck_ballots';

function hackyTallyVotesDoNotUseThis(
  gridPositions: GridPosition[]
): Map<ContestId, Map<CandidateId, number>> {
  const tally = new Map<ContestId, Map<CandidateId, number>>();

  for (const gridPosition of gridPositions) {
    const { contestId } = gridPosition;

    if (!tally.has(contestId)) {
      tally.set(contestId, new Map());
    }

    const contestTally = tally.get(contestId)!;
    const optionId =
      gridPosition.type === 'option'
        ? gridPosition.optionId
        : `write-in-${gridPosition.writeInIndex}`;

    contestTally.set(optionId, (contestTally.get(optionId) ?? 0) + 1);
  }

  return tally;
}

test('generateTestDeckBallots', () => {
  const { election } = electionGridLayoutNewHampshireHudsonFixtures;

  for (const ballotStyle of election.ballotStyles) {
    const testBallots = generateHandMarkedTestDeckBallots({
      election,
      ballotStyleId: ballotStyle.id,
      includeOvervotedBallots: false,
    });

    expect(
      hackyTallyVotesDoNotUseThis(
        testBallots.flatMap(({ gridPositions }) => gridPositions)
      )
    ).toMatchSnapshot();
  }
});

test('generateTestDeckBallots with overvotes', () => {
  const { election } = electionGridLayoutNewHampshireHudsonFixtures;

  for (const ballotStyle of election.ballotStyles) {
    const testBallots = generateHandMarkedTestDeckBallots({
      election,
      ballotStyleId: ballotStyle.id,
      includeOvervotedBallots: true,
    });

    expect(
      hackyTallyVotesDoNotUseThis(
        testBallots.flatMap(({ gridPositions }) => gridPositions)
      )
    ).toMatchSnapshot();
  }
});

test('invalid ballot style', () => {
  const { election } = electionGridLayoutNewHampshireHudsonFixtures;

  expect(() =>
    generateHandMarkedTestDeckBallots({
      election,
      ballotStyleId: 'invalid',
    })
  ).toThrowErrorMatchingInlineSnapshot(`"Ballot style not found: invalid"`);
});
