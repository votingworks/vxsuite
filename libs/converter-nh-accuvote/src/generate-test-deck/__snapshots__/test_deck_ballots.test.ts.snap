// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateTestDeckBallots 1`] = `
Map {
  "President-and-Vice-President-of-the-United-States-63e49f46" => Map {
    "Donald-J-Trump-and-Michael-R-Pence-3e3f31a7" => 1,
    "Joseph-R-Biden-and-Kamala-D-Harris-346756c0" => 1,
    "Jo-Jorgensen-and-Jeremy-Cohen-7e981234" => 1,
    "write-in-0" => 1,
  },
  "Governor-9e42d79b" => Map {
    "Chris-Sununu-130124f7" => 1,
    "Dan-Feltes-3fadf798" => 1,
    "Darryl-W-Perry-2f07c455" => 1,
    "write-in-0" => 1,
  },
  "United-States-Senator-19e5c11e" => Map {
    "Corky-Messner-b6eb216d" => 1,
    "Jeanne-Shaheen-6e19622d" => 1,
    "Justin-O-Donnell-ec71760c" => 1,
    "write-in-0" => 1,
  },
  "Representative-in-Congress-838abf88" => Map {
    "Steven-Negron-5d482d72" => 1,
    "Ann-McLane-Kuster-67fdf060" => 1,
    "Andrew-Olding-bed33f08" => 1,
    "write-in-0" => 1,
  },
  "Executive-Councilor-79d43428" => Map {
    "Dave-Wheeler-de2242ee" => 1,
    "Debora-B-Pignatelli-8c4cc1d3" => 1,
    "write-in-0" => 1,
  },
  "State-Senator-4ba324f1" => Map {
    "Sharon-M-Carson-0316308e" => 1,
    "Nancy-Hendricks-d330559d" => 1,
    "write-in-0" => 1,
  },
  "State-Representatives-ef8e6428" => Map {
    "Tony-Lekas-a06f7364" => 1,
    "Hershel-Nunez-6862315f" => 1,
    "Lynne-Ober-f704df17" => 1,
    "Russell-Ober-e433afea" => 1,
    "Andrew-Prout-1cd243e0" => 1,
    "Andrew-Renzullo-b0428370" => 1,
    "Kimberly-Rice-faaed16a" => 1,
    "Denise-Smith-0a6359c3" => 1,
    "Jordan-Ulery-28553627" => 1,
    "Bob-Greene-c006a3bd" => 1,
    "Alicia-Lekas-271048d8" => 1,
    "Brett-Gagnon-7a8c9838" => 1,
    "David-Hennessey-f1343f4f" => 1,
    "Beatriz-Jauregui-11b83653" => 1,
    "Steven-Katsos-077584b8" => 1,
    "Harold-Lynde-a54df7d7" => 1,
    "Lana-Paliy-595b276a" => 1,
    "Robert-S-Sherman-f10f4a3f" => 1,
    "Alejandro-Urrutia-1b574b86" => 1,
    "Timothy-Wyatt-0105b8de" => 1,
    "Barbara-A-Blue-14608463" => 1,
    "Nancy-S-Brucker-4471faf7" => 1,
    "write-in-0" => 1,
    "write-in-1" => 1,
    "write-in-2" => 1,
    "write-in-3" => 1,
    "write-in-4" => 1,
    "write-in-5" => 1,
    "write-in-6" => 1,
    "write-in-7" => 1,
    "write-in-8" => 1,
    "write-in-9" => 1,
    "write-in-10" => 1,
  },
  "Sheriff-ee138ee1" => Map {
    "Christopher-Connelly-970b2364" => 1,
    "Bill-Barry-0e677ca6" => 1,
    "write-in-0" => 1,
  },
  "County-Attorney-3f2a6707" => Map {
    "John-J-Coughlin-0e7bf93c" => 1,
    "Michael-Conlon-1a61fa16" => 1,
    "Nicholas-Sarwark-c8f115e2" => 1,
    "write-in-0" => 1,
  },
  "County-Treasurer-bd2e6c39" => Map {
    "David-G-Fredette-687f01ae" => 1,
    "William-Bryk-eaf936ec" => 1,
    "Richard-Manzo-c5fcc876" => 1,
    "write-in-0" => 1,
  },
  "Register-of-Deeds-75774240" => Map {
    "Jack-Flanagan-78e24cab" => 1,
    "Mary-Ann-Crowell-e1ff59b9" => 1,
    "write-in-0" => 1,
  },
  "Register-of-Probate-dd373952" => Map {
    "Elizabeth-Ann-Moreau-aec07bb4" => 1,
    "Elizabeth-Ropp-4035eed1" => 1,
    "write-in-0" => 1,
  },
  "County-Commissioner-72fa4aa2" => Map {
    "Michael-Soucy-b23f7c2d" => 1,
    "Daniel-C-Hansberry-8c0291ec" => 1,
    "write-in-0" => 1,
  },
}
`;
